sound/soc/sof/pcm.c

```diff
@@ -477,13 +477,9 @@ static int sof_pcm_open(struct snd_soc_component *component,
 
 	caps = &spcm->pcm.caps[substream->stream];
 
-	/* set any runtime constraints based on topology */
-	snd_pcm_hw_constraint_step(substream->runtime, 0,
-				   SNDRV_PCM_HW_PARAM_BUFFER_BYTES,
-				   le32_to_cpu(caps->period_size_min));
-	snd_pcm_hw_constraint_step(substream->runtime, 0,
-				   SNDRV_PCM_HW_PARAM_PERIOD_BYTES,
-				   le32_to_cpu(caps->period_size_min));
+	/* set runtime constraint to use integer period count */
+	snd_pcm_hw_constraint_integer(substream->runtime,
+				      SNDRV_PCM_HW_PARAM_PERIODS);
```

Comment:
why would you remove the previous constraints? Couldn't you just *ADD* a new constraint that requires both that the periods and buffers are larger than a min, and have an integer relationship.

What test exposed this and what is this about really?

Reply:
also your commit message is particularly unclear. You state "we already have rule to make sure that the buffer_size will be the "period_size * periods" in ALSA core," so how could the relationship NOT be integer? Or are you saying that we need to apply a pre-defined rule provided by the ALSA core?

Reply:
can someone give me an example of a problem with the existing code? in what cases do we have a buffer/period that is NOT a multiple of the min period?

I am not against change, I'd like to know if this is a real problem or a hypothetical one. If it ain't broke don't fix it...

Reply:
There was a historical constraint on the FW host DMA where host buffer/period size had to be divisible by FW period size, but I believe that no longer applies.

Reply:
> @plbossart @keyonjie I think the problem with old code is that we align both buffer and period sizes to be multiples of caps->period_size_min. This seems too constraining and will causes failures with params we really ought to support just fine. We already declare minmax in sof_pcm_open(), so the above constraint should be good enough.
> 
> So while the commit message can be clarified, I think the change is good.

Thanks, @kv2019i you get it exactly, let me try to clarify more in the commit message.

Reply:
> There was a historical constraint on the FW host DMA where host buffer/period size had to be divisible by FW period size, but I believe that no longer applies.

Yes, we had that, but even if that is even applies, we should change to periods be integer which is the correct one as done in my PR.

Reply:
I am not fully on board with this change.
The min period size typically defines the size of a DMA transfer, and typically defined the 'rewind safeguard' for PulseAudio. 
If we allow for periods/buffers that are not aligned with the elementary DMA transfers, then we are signing up for trouble. I just don't understand how the hardware would deal with this.

And I must admit I don't understand your commit message
"with example period_size_min of 192, period size is constrained to be
192*n, and we can't e.g. set the period_bytes to be 256bytes, even
though it is in the range."

What prevents you from setting a min_period_size of 256, if this is what you want and it's supported by the firmware?

I really feel like you are removing a valid check when you should be fixing the topology file to use min_period_size values that make sense, instead of copy-pasting 192.

Reply:
@keyonjie please address this feedback

Reply:
@plbossart This is a bit mixed messaging as @lgirdwood said earlier this is _not_ the minimum DMA transfer size for FW, but you are saying that it is. To me 192 sounds like an odd transfer size alignment constraint for DMA, but if it is, then I stand corrected and then this patch may need to be revisited. 

For the commit message, I think it makes sense. Changing topology is not really something the end user (or even kernel developer can do). So I think the commit message is right, you cannot set period_size to 256 from the application as all our topologies have alignment to 192 bytes.

Reply:
If the firmware starts pulling data over DMA, do we agree yes/no that that amount of data is no longer accessible by the host?
All in-flight data defines the minimum period size, and that period may be larger due to hardware alignment.
I am not sure why we should remove restrictions to accommodate bad values in the topology.

Reply:
In the case of HDaudio DMA, I believe the minimum period size is defined by the DMA alignment requirements (3.3.32 in the HDaudio spec), so 128 bytes

Since the min period size if defined by topology, then currently the topology needs to be corrected to account for this restriction.

Alternatively, maybe we should have platform-specific fields for min period sizes that are set in the driver. I agree that maybe those values belong in drivers and not topology.

Reply:
@plbossart as @kv2019i mentioned, the caps->period_size_min is not the minimum DMA transfer size, it has no relationship with any platform specific DMA alignment requirements. Instead, it was set to 192Bytes (1ms data, S16, 2Bytes * 2 channel * 48) in almost all topology files, as @lgirdwood mentioned, it was set to try to make sure the host/ALSA buffer size be multiple times of the FW host buffer period size(1ms typically).

IIRC, the minimum DMA transfer size for DW DMA can be as smaller as 2Bytes or at least 4Bytes, so we don't have to restrict the ALSA buffer to be integer ms size here in the SOF generic PCM driver, considering the 192Bytes can't make sure it is integer ms size for some other format(e.g. S24_3LE).

To be simple, the caps->period_size_min from topology today doesn't have any meaning, we should not apply any constraints based on it, so we should either change it(to be 1?) in all those topology files or remove using it in the driver as the change I made here.

Reply:
I don't mind if you remove this check on topology values, but we have to keep a hardware-based constraint. Your answer just shows that we cannot use the same values for BYT/BDW and SKL+. So my suggestion is to add period_min in the platform-specific drivers, and to remove the topology values. Or use a max(hardware value, topology value).

Reply:
> I don't mind if you remove this check on topology values, but we have to keep a hardware-based constraint. Your answer just shows that we cannot use the same values for BYT/BDW and SKL+. So my suggestion is to add period_min in the platform-specific drivers, and to remove the topology values. Or use a max(hardware value, topology value).

Sorry, I don't know what platform-specific period_min values we should set for BYT/BDW and SKL+, that should be another fix(comparing to my removal of the incorrect constraints here) no?

Reply:
No. I don't want to remove constraints on period_min, if the topology values are incorrect, then we need to have platform-specific ones.

Just take the values used for legacy:

````
atom/sst-mfld-platform-pcm.c:   .periods_min = SST_MIN_PERIODS,
atom/sst-mfld-platform-pcm.c:   .periods_max = SST_MAX_PERIODS,
baytrail/sst-baytrail-pcm.c:    .periods_min            = 2,
baytrail/sst-baytrail-pcm.c:    .periods_max            = 250,
haswell/sst-haswell-pcm.c:      .periods_min            = HSW_PCM_PERIODS_MIN,
haswell/sst-haswell-pcm.c:      .periods_max            = HSW_PCM_PERIODS_MAX,
skylake/skl-pcm.c:      .periods_min =          2,
skylake/skl-pcm.c:      .periods_max =          AZX_MAX_FRAG,

atom/sst-mfld-platform-pcm.c:   .period_bytes_min = SST_MIN_PERIOD_BYTES,
atom/sst-mfld-platform-pcm.c:   .period_bytes_max = SST_MAX_PERIOD_BYTES,
baytrail/sst-baytrail-pcm.c:    .period_bytes_min       = 384,
baytrail/sst-baytrail-pcm.c:    .period_bytes_max       = 48000,
haswell/sst-haswell-pcm.c:      .period_bytes_min       = PAGE_SIZE,
haswell/sst-haswell-pcm.c:      .period_bytes_max       = (HSW_PCM_PERIODS_MAX / HSW_PCM_PERIODS_MIN) * PAGE_SIZE,
skylake/skl-pcm.c:      .period_bytes_min =     128,
skylake/skl-pcm.c:      .period_bytes_max =     AZX_MAX_BUF_SIZE / 2,

````

Reply:
Thanks @plbossart and @keyonjie for clarifying! So with that, my thumbs up to use the legacy values used by SST and have them in driver code.

Reply:
@dbaluta FYI, you probably also need to look for relevant min period/buffer values for i.MX

Reply:
@plbossart I believe those ones for legacy platforms(e.g. 128 for skl, 384 for byt, 4096 for hsw/bdw) are not related with the minimum DMA transfer size, so I would suggest not to copy those values for SOF.

On the other hand, I believe the change I am making here is not mandatory for KWD (it is actually fix for the SOF common issue that I had described), I am suggesting the Chrome team to go without this change and will close the PR here once it is verified work like that.

Reply:
so we've discussed a PR that wasn't necessary, and we are not going to fix what is obviously broken (topology-defined min sizes).

@keyonjie please be mindful of reviewer time, you have multiple PRs in flight and every time you abandon them for your future contributions.

Reply:
@plbossart my point is that the PR is necessary but *not* *mandatory* for S0ix KWD fix, I submitted the PR as I have been thinking it is helpful for the SOF common driver(to make the topology-configured buffer-size range full usable and be friendly to user apps), I didn't intend to waste reviewers' time, sorry if that has been the fact.